Adds support for detaching sources from customers

ob-stripe opened the pull request:
r? @brandur-stripe 
cc @stripe/api-libraries @will-stripe

This PR adds support for detaching sources from customers. Unlike cards or bank accounts, sources cannot be deleted, but they can now be detached from customers by sending a `DELETE` request to `/v1/customers/{CUSTOMER_ID}/sources/{SOURCE_ID}`.

Incidentally, while working on this PR, I stumbled upon a couple of related issues:

- we were lacking an entry for `source` in  [`Util.object_classes`](https://github.com/stripe/stripe-ruby/blob/5dd5d2b608af8b597b5298b22eab171d570d6408/lib/stripe/util.rb#L20). This means that `source` objects returned by the API were instantiated as generic `StripeObject`s rather than `Source`s

- the `source` entry in `fixtures.json` currently contains a card object rather than a source object

  I've fixed the source tests to (correctly) expect a source object, which means tests won't pass until `fixtures.json` is fixed and this PR is rebased.

brandur-stripe commented:
Sources are fixed:

https://github.com/stripe/openapi/blob/master/spec/fixtures.yaml#L862-L894

@ob-stripe Can you pull down the new spec and add it to this branch? I'd put it in master, but we need your changes to `util.rb` for source objects to be recognized. Instructions on how to bump the OpenAPI spec are in the README.

You might have to make these changes as well to get the test suite fully passing (because the source fixture changed its type):

``` patch
diff --git a/test/stripe/source_test.rb b/test/stripe/source_test.rb
index 6c66e6c..6c662a8 100644
--- a/test/stripe/source_test.rb
+++ b/test/stripe/source_test.rb
@@ -16,7 +16,7 @@ module Stripe
         token: API_FIXTURES.fetch(:token)[:id]
       )
       assert_requested :post, "#{Stripe.api_base}/v1/sources"
-      assert source.kind_of?(Stripe::Card)
+      assert source.kind_of?(Stripe::Source)
     end

     should "be saveable" do
@@ -29,7 +29,7 @@ module Stripe
     should "be updateable" do
       source = Stripe::Source.update(FIXTURE[:id], metadata: {foo: 'bar'})
       assert_requested :post, "#{Stripe.api_base}/v1/sources/#{FIXTURE[:id]}"
-      assert source.kind_of?(Stripe::Card)
+      assert source.kind_of?(Stripe::Source)
     end

     context "#verify" do
```

Thanks!

ob-stripe commented:
@brandur-stripe Okay, I updated the `fixtures.json` / `.yaml` files, which revealed another couple of issues:

- the `Card` tests were using the `"source"` fixture rather than the `"card"`

- the fixture objects include keys with empty strings as their values (e.g. `"customer": "",`) when in reality the API would omit the key entirely

Previously, cards were simply testing the presence of the `account` / `customer` / `recipient` keys (via `respond_to?`) to determine which resource they belonged to, which in turn determines the URL for update & delete requests.

I fixed the issue by making the test more thorough: now, cards not only test that each key exists, but also that it has a non-null and non-empty value.

I also updated sources to do the same thing. I did so in separate commits to make the PR easier to review, but let me know if you'd rather tackle each issue through separate PRs.

brandur-stripe commented:
Nice!

> the Card tests were using the "source" fixture rather than the "card"

Oops, that's my bad! Thanks for the fix.

> the fixture objects include keys with empty strings as their values (e.g. "customer": "",) when in reality the API would omit the key entirely

Arg, yeah. I think that's an unfortunate quirk in the data generator, and it would be nice to fix it. For now though, it probably makes sense to just tolerate a number of different "known empty conditions" especially given that the nature of the library is so dynamic.

> I fixed the issue by making the test more thorough: now, cards not only test that each key exists, but also that it has a non-null and non-empty value.

Yep, this makes sense to me. I had to do this somewhere else in the code at some point too. I'm kind of thinking that maybe we should even just have a `Utils.present?(...)` type function that just encapsulates all that logic to make safety there more convenient and hopefully widespread.

This looks good for now though. Pulling it in!

brandur-stripe commented:
Released as 2.1.0.
